HPCC-29334 Associate log events with active Trace/Span

rpastrana opened the pull request:
- Adds Trace and Span ID columns to JLOG
- Adds TraceInfo Thread tracking logic
- Automatically assigns threadwise traceinfo
- Adds manual unittest

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

rpastrana committed:
HPCC-29334 add traceid to log events

- Adds Trace and Span ID columns to JLOG
- Adds TraceInfo Thread tracking logic
- Automatically assigns threadwise traceinfo
- Adds manual unittest

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

rpastrana committed:
HPCC-29334 Testing serialization of trace info

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

ghalliday reviewed (commented):
@rpastrana the bulk of the changes look good.  A couple of things need some discussion:
- serialization, especially backward compatibility
- exactly how the active trace is maintained per thread.  I think this may need some refactoring.

rpastrana committed:
HPCC-29334 Code review 2

 - Removes use of unknownTrace.queryTraceID
 - Removes de/seriliazation logic
 - No longer adds/removes traceinfo via logmsg push
 - Adds several constructors to avoid spread out logic
 - Does not enter criticalblock if lookup value is invalid
 -

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

ghalliday commented on system/jlib/jlog.cpp:
```diff
@@ -292,28 +293,33 @@ void LogMsgJobInfo::deserialize(MemoryBuffer & in)
 }
 
 static LogMsgJobInfo globalDefaultJobInfo(UnknownJob, UnknownUser);
+static LogMsgTraceInfoId globalDefaultTraceInfoId = UnknownTraceInfoId;
```
Not actually used and can be deleted (I will; eventually do the same for the glolbalDefaultJobInfo

ghalliday commented on system/jlib/jlog.cpp:
```diff
@@ -292,28 +293,33 @@ void LogMsgJobInfo::deserialize(MemoryBuffer & in)
     // There is an alternative interpretation, that an unset thread-local one should default to whatever the global one is at the time the thread one is used.
     // In practice I doubt there's a lot of difference as global one is likely to be set once at program startup
     defaultJobInfo = globalDefaultJobInfo;
+    setDefaultTraceInfo(_traceInfoId, true);
```
probably easier to undestand setting
```
 defaultTraceInfoId = _traceInfoId;
```
Since this is an internal restoring operation.

ghalliday commented on system/jlib/jlog.cpp:
```diff
@@ -327,6 +333,32 @@ const LogMsgJobInfo & checkDefaultJobInfo(const LogMsgJobInfo & _jobInfo)
+    return setDefaultTraceInfo(theManager->addTraceInfo(theTraceID, theSpanID), threaded);
+}
+
+LogMsgTraceInfoId setDefaultTraceInfo(LogMsgTraceInfoId id, bool threaded)
```
cleanup: The threaded parameter can be removed since globalDefaultTraceInfoId is not needed.

ghalliday commented on system/jlib/jlog.cpp:
```diff
@@ -327,6 +333,32 @@ const LogMsgJobInfo & checkDefaultJobInfo(const LogMsgJobInfo & _jobInfo)
+
+LogMsgTraceInfoId setDefaultTraceInfo(const char *theTraceID, const char *theSpanID, bool threaded)
+{
+    return setDefaultTraceInfo(theManager->addTraceInfo(theTraceID, theSpanID), threaded);
```
The code would be cleaner/more symmetric if this addTraceInfo() call was in the CSpan (to match the remove call).  However if everything turns out cleaner storing an ISpan in the thread context then that will remove the need for this add.

ghalliday reviewed (commented):
@rpastrana a few comments.  I think there is also an issue with the lifetime of the ids if the logging is output on a different thread.  It may be worth exploring if saving the ISpan works better before delving any further.
See https://github.com/hpcc-systems/HPCC-Platform/pull/18305 for a PR to clean up the context saving/restoring.

ghalliday reviewed (commented):
removing my need to review until this has been revisited.
